### DISCUSSION TITLE
How to debug Nuxt 3 in VSCODE

### DISCUSSION THREAD

**Discussion by igeniusgroup** [nuxt/nuxt#15964] [General]
Hi Nuxt team,

Can you help me with some instructions on how to debug NUXT 3 in VSCODE... I'm pulling my hair out trying to configure it and we want to use NUXT 3 for our next project.

Thanks in advance for your thoughts and help :-)

AJ

**Comment by kissu** [comment]
How to debug what? This is a pretty large question.
If you're speaking of devtools, be sure that you do have the [Vue3 ones](https://chrome.google.com/webstore/detail/vuejs-devtools/ljjemllljcmogpfapbkkighbhhppjdbg). It will help you pretty well already.
Then, if you have a more specific question, feel free to update your question with more details! 👍🏼

**Reply by wzdxy** [reply]
I think @igeniusgroup  means to debug the server side code. 
For example, the ts files in server，and js code in vue files when it exec in server side lifecycle).
I need it too.

**Reply by dacarey** [reply]
As @wzdxy  says,  it is server-side as well as client-side that clear instructions would be useful for for Nuxt 3

There are some great instructions and tutorials for Nuxt 2.  Here is one such tutorial.   
However following the same for Nuxt 3 and I'm unable (yet) to step through server-side code

Any insights greatly appreciated!

**Comment by AXSJ** [comment]
Hello everyone. Is there a reason we haven't received a response for this? This is a vital thing. We need to be able to debug Nuxt 3

**Reply by twofingerrightclick** [reply]
I agree. I have found the Nuxt 3 docs decently complete in regards to how to program, but very lacking in details about setting up the tooling.  I really hope some people author some pages about tooling setup geared to an audience other than experienced Typescript tooling web devs. This issue of missing debugging docs is like step 2 in app programming and yet there is no table of contents heading for the topic in Nuxt 2 or 3 docs. Just a generic overview of course is all that is desired. However, Nuxt 3 has been so easy to get going with compared to Nuxt 2! I have spent so much less time with the tooling.

**Comment by syavorsky** [comment]
Landed on this page asking the same question. In case someone else still needs it, this worked for me on the server-side. 
This VSCode config with explicit `debugger` statement in the code. 

```json
{
      "name": "Run and debug",
      "type": "node",
      "request": "launch",
      "cwd": "${workspaceFolder}",
      "runtimeExecutable": "npm",
      "runtimeArgs": ["run", "dev"]
} 
```

**Reply by mihail727** [reply]
thx that realy works

**Reply by 60pfennig** [reply]
thanks :*
but how to debug? With your config the debugger starts but I only get unbound breakpoints.
Edit: Debugging Client side code was the key
https://nordschool.com/enable-vs-code-debugger-for-nuxt-and-typescript/

**Reply by Maxim-Filimonov** [reply]
I don't think it works with vscode debug breakpoints but putting `debugger` in your code does work.
Confusingly, it would stop in your mjs file instead of source mapped code. Not the best experience but better than nothing

**Comment by tobiasdiez** [comment]
A simple way to debug the server code is starting the dev server in the [Javascript debug console](https://code.visualstudio.com/docs/nodejs/nodejs-debugging#_javascript-debug-terminal). Everything then just seems to work automatically, including breakpoints and source maps.

**Comment by ckhatton** [comment]
Using @tobiasdiez's advice (thank you!), you can create these two files to do the same...

`/.vscode/launch.json`
```json
{
  // Use IntelliSense to learn about possible attributes.
  // Hover to view descriptions of existing attributes.
  // For more information, visit: https://go.microsoft.com/fwlink/?linkid=830387
  "version": "0.2.0",
  "compounds": [
    {
      "name": "fullstack: Nuxt & Chrome",
      "configurations": ["server: Nuxt", "client: Chrome"]
    }
  ],
  "configurations": [
    {
      "name": "client: Chrome",
      "request": "launch",
      "type": "chrome",
      "url": "http://localhost:3000",
      "webRoot": "${workspaceFolder}",
      "preLaunchTask": "Sleep Delay"
    },
    {
      "name": "server: Nuxt",
      "type": "node",
      "request": "launch",
      "cwd": "${workspaceFolder}",
      "runtimeExecutable": "npm",
      "runtimeArgs": ["run", "dev"],
      "console": "integratedTerminal"
    }
  ]
}
```

`/.vscode/tasks.json`
```json
{
  // See https://go.microsoft.com/fwlink/?LinkId=733558
  // for the documentation about the tasks.json format
  "version": "2.0.0",
  "tasks": [
    {
      "label": "Sleep Delay",
      "type": "shell",
      "command": "sleep 5",
      "presentation": {
        "reveal": "never",
        "panel": "shared",
        "close": true
      }
    }
  ]
}
```

> **Note:**
> Adjust the browser choice and sleep time to suit your needs and system speed, respectively.

This will add an option labelled "fullstack: Nuxt & Chrome", which will run the whole lot. You can still run "server: Nuxt" and "client: Chrome" independently, but the browser launch will be delayed due to the sleep script - so (`cmd`/`super`/`win`)+`click` the URL in the terminal instead.

:octocat:

**Reply by CharlesBT** [reply]
Hi, to amend the @ckhatton config, I use the following config :
- no need to sleep :) : use the problem matcher to launch browser when Nitro is ready, matching output console for "√ Nitro built in"
- automatic launch of your custom Chrome config, including extension of your choice like Vue Devtools

`/.vscode/launch.json`
```json
{
  // Use IntelliSense to learn about possible attributes.
  // Hover to view descriptions of existing attributes.
  // For more information, visit: https://go.microsoft.com/fwlink/?linkid=830387
  "version": "0.2.0",
  "configurations": [
    // launch config for debugging Vue web apps
    {
      "name": "Debug Nuxt: WebApp",
      "type": "chrome",
      "presentation": {
        "hidden": false,
        "group": "1 PROGRAM:NUXT",
        "order": 3
      },
      "request": "launch",
      "url": "http://localhost:3000",
      "webRoot": "${workspaceFolder}",
      "preLaunchTask": "nuxt:dev",
      "userDataDir": "${workspaceFolder}/.vscode/vscode-chrome-debug-userdatadir"
      // "userDataDir": "${env:HOME}/.vscode/vscode-chrome-debug-userdatadir"
      // This config is to always use the same configuration for the chrome app (system wide)
      // therefore you can install the prefered extensions once and they will stay!
      // Taken from here: https://stackoverflow.com/questions/51725854/run-vscode-chrome-debugger-with-its-extensions
    },

    {
      "name": "Debug Nuxt: Server",
      "type": "node",
      "presentation": {
        "hidden": false,
        "group": "1 PROGRAM:NUXT",
        "order": 2
      },
      "request": "launch",
      "cwd": "${workspaceFolder}",
      "runtimeExecutable": "pnpm",
      "windows": {
        "runtimeExecutable": "pnpm.cmd"
      },
      "runtimeArgs": ["dev"]
    }
  ],
  "compounds": [
    {
      "name": "Debug Nuxt: Server & WebApp",
      "presentation": {
        "hidden": false,
        "group": "1 PROGRAM:NUXT",
        "order": 1
      },
      "stopAll": true,
      "configurations": ["Debug Nuxt: Server", "Debug Nuxt: WebApp"]
    }
  ]
}

```


`/.vscode/tasks.json`
```json
{
  "version": "2.0.0",
  "tasks": [
    {
      "label": "nuxt:dev",
      "type": "npm",
      "script": "dev",
      "isBackground": true,
      "problemMatcher": [
        {
          "base": "$tsc-watch",
          "background": {
            "activeOnStart": true,
            "beginsPattern": "√ Nitro",
            "endsPattern": "built in"
          }
        }
      ],
      "group": {
        "kind": "build",
        "isDefault": true
      }
    }
  ]
}
```

**Reply by jayEvans89** [reply]
Using the above I'm still getting unbound breakpoints on any server code. Anyone got any pointers on what direction to go about getting it working?

**Reply by ckhatton** [reply]
Nice one @CharlesBT! 👌

**Reply by ckhatton** [reply]
Yeh, unfortunately debugging Nuxt still seems to be temperamental @jayEvans89. Try placing the word `debugger` where you want a breakpoint and see if that helps.

**Comment by green-anger** [comment]
This worked for me - [official docs](https://nuxt.com/docs/guide/going-further/debugging#example-vs-code-debug-configuration). [Sourcemaps](https://nuxt.com/docs/guide/going-further/debugging#sourcemaps) are activated by default:

> Sourcemaps are enabled for your server build by default, and for the client build in dev mode...

Breakpoints triggered in both client and server.

**Reply by knunn-kygunco** [reply]
Like you, I copied exactly what was in the documentation. When I run the "client: chrome" debugger, it says localhost is refusing to connect. Any idea why this might be?

**Comment by x11joe** [comment]
I wanted to mention if you are on docker, and using remote shell to login to the container to use debugging make sure to add

export EDITOR=code in your package.json

so like this

"scripts": {
    "build": "nuxt build",
    "dev": "export EDITOR=code && nuxt dev",
    "start": "nuxt start",
    "generate": "nuxt generate",
    "preview": "nuxt preview",
    "postinstall": "nuxt prepare"
  },
  
  This will fix the error where it can't load the vscode editor because it's trying to use vim instead

**Comment by gorrus100** [comment]
https://nuxt.com/docs/guide/going-further/debugging#example-vs-code-debug-configuration

select mode 'fullstack'
working debug with client and server side

**Comment by codywakeford** [comment]
Do watch statements work too? I cant get them to work in vsccode.
